File: events/qconsp2017/code/sample-1/Req.java
//Sample provided by Fabio Galuppo  
//April 2017

//compile:
//sudo javac -d bin -cp .:../refs/zmq.jar Req.java

//run:
//java -Djava.library.path="/Users/fabiogaluppo/zmq/samples/refs" -cp .:./bin:../refs/zmq.jar Req

import org.zeromq.ZMQ;
import org.zeromq.ZMQ.Socket;
import org.zeromq.ZMQ.Context;

public final class Req {
    @SuppressWarnings("unchecked")
    public static void main(String[] args) {
        System.out.println("ZMQ version = " + ZMQ.getVersionString());

        Context zmqContext = ZMQ.context(1);

        Socket requestSocket = zmqContext.socket(ZMQ.REQ);
        requestSocket.connect("tcp://localhost:60000");

        String msg = "Hello, World!";
        if (args.length > 0)
            msg = args[0];
        
        requestSocket.send(msg);
        byte[] data = requestSocket.recv(0);
        String text = new String(data);
        System.out.println(text);

        requestSocket.close();
        zmqContext.close();
    }
} 

File: events/ccppbrasil_13o_encontro/code/sample-2/req.cpp
//Sample provided by Fabio Galuppo  
//August 2017

//compile:
//clang++ -O3 -std=c++14 -I../include -I./cppzmq/include -L/usr/local/Cellar/zeromq/4.2.2/lib -lzmq req.cpp -o ./bin/req.exe

//run:
//./bin/req.exe

#include <zmq.hpp>

#include <string>
#include <iostream>
#include <cstring>

#include "zmq_utils.hpp" 

int main(int argc, char* argv[]) {
    std::cout << "ZMQ version = " << zmq_utils::get_zmq_version() << "\n";
    
    zmq::context_t zmq_context(1);
    zmq::socket_t request_socket(zmq_context, ZMQ_REQ);

    request_socket.connect("tcp://localhost:60000");
    
    const char* msg = "Hello, World!";
    if (argc > 1)
        msg = argv[1];
 
    zmq::message_t zmsg{msg, msg + strlen(msg)};
    request_socket.send(zmsg, 0);

    request_socket.recv(&zmsg, 0);
    size_t size = zmsg.size();
    if (size > 0) {
        const char* begin = static_cast<const char*>(zmsg.data());
        const char* end = begin + size;
        std::cout << std::string(begin, end) << "\n";
    } else {
        std::cout << "no data returned from recv\n";
    }
    
    request_socket.close();
    zmq_context.close();

    return 0;
}